src/main/java/org/truffleruby/core/kernel/KernelNodes.java

```diff
@@ -664,6 +657,32 @@ protected boolean assignsNoNewVariables(RootNodeWrapper rootNode) {
         private boolean frameHasOnlySelf(final FrameDescriptor descriptor) {
             return descriptor.getSize() == 1 && SelfNode.SELF_IDENTIFIER.equals(descriptor.getSlots().get(0).getIdentifier());
         }
+
+        @TruffleBoundary
+        public static String offsetSource(String method, String source, String file, int line) {
+            // TODO CS 23-Apr-18 Truffle doesn't support line numbers starting at anything but 1
+            if (line == 0) {
+                // fine instead of warning because these seem common
+                Log.LOGGER.fine(() -> String.format("zero line number %s:%d not supported in #%s - will be reported as starting at 1", file, line, method));
+                return source;
+            } else if (line < 1) {
+                Log.LOGGER.warning(String.format("negative line number %s:%d not supported in #%s - will be reported as starting at 1", file, line, method));
+                return source;
+            } else if (line > 1) {
+                // fine instead of warning because we can simulate these
+                Log.LOGGER.fine(() -> String.format("offset line number %s:%d are simulated in #%s by adding blank lines", file, line, method));
+
+                final StringBuilder builder = new StringBuilder();
+                for (int n = 1; n < line; n++) {
+                    builder.append("\n");
+                }
```

Comment:
This seems fairly inefficient.
I know @nirvdrum optimized this, so please use the code below:
```java
            final char[] emptyLines = new char[Math.max(line - 1, 0)];
            Arrays.fill(emptyLines, '\n');
            Source source = Source.newBuilder(new String(emptyLines) + code)
```

Reply:
Ok.